Add Tutorial Website (#35)
ad3154/ISO11783-CAN-Stack: sphinx/source/Tutorials/The ISOBUS Hello World.rst

.. _ISOBUSHellowWorld:

The ISOBUS Hello World!
========================

.. toctree::
   :hidden:
   :glob:  

.. contents:: Contents
   :depth: 2
   :local:

In this tutorial, we'll demonstrate how to set up the library in a brand new C++ program and send a basic message to the broadcast address!

Let's get started by setting up a new C++ project to use this library.

We'll assume you followed the instructions in the :doc:`installation section <../Installation>`, so if you haven't, be sure to go back and check those instructions out.

> We'll also assume you've reviewd the :doc:`concepts section <../Concepts>`, and are famillier with things like a NAME, and the idea of a control function.

Reviewer: ```suggestion
We'll also assume you've reviewed the :doc:`concepts section <../Concepts>`, and are familiar with things like a NAME, and the idea of a control function.
```

------------

Let's set up a standard, empty C++ program in a file called "main.cpp", and include the basic header files we'll need from the library to start using it.

.. code-block:: c++

   #include "can_network_manager.hpp"
   #include "socket_can_interface.hpp"
   #include "can_partnered_control_function.hpp"
   
   int main()
   {
   	return 0;
   }

The file `"can_network_manager.hpp" <https://delgrossoengineering.com/isobus-docs/can__network__manager_8hpp.html>`_ is the header file for the main interface for the CAN stack.

The file `"socket_can_interface.hpp" <https://github.com/ad3154/ISO11783-CAN-Stack/blob/main/socket_can/include/socket_can_interface.hpp>`_ is a hardware tie-in for socket CAN.

The file `"can_partnered_control_function.hpp" <https://delgrossoengineering.com/isobus-docs/can__partnered__control__function_8hpp.html>`_ Defines the interface for a *partnered control function*, which is another device you want to communicate with.

Tell the stack who you are (NAME)
----------------------------------

Before you can communicate on the CAN bus, you need to come up with a `NAME <https://delgrossoengineering.com/isobus-docs/classisobus_1_1NAME.html>`_.

> A NAME is what uniquely identifes your device on the bus.

Reviewer: ```suggestion
A NAME is what uniquely identifies your device on the bus.
```

Check out the :doc:`concepts section <../Concepts>`, if you need to brush up on NAMEs.

Let's create a `NAME <https://delgrossoengineering.com/isobus-docs/classisobus_1_1NAME.html>`_ in our program.

> NOTE: Eveything that is part of the stack is in the namespace `"isobus" <https://delgrossoengineering.com/isobus-docs/namespaceisobus.html>`_!

Reviewer: ```suggestion
NOTE: Everything that is part of the stack is in the namespace `"isobus" <https://delgrossoengineering.com/isobus-docs/namespaceisobus.html>`_!
```

.. code-block:: c++

   #include "can_network_manager.hpp"
   #include "socket_can_interface.hpp"
   #include "can_partnered_control_function.hpp"
   
   int main()
   {
    isobus::NAME myNAME(0); // Create an empty NAME

    return 0;
   }

Now we have a NAME instantiated. We still need to populate it with our control function's details.

.. code-block:: c++

   #include "can_network_manager.hpp"
   #include "socket_can_interface.hpp"
   #include "can_partnered_control_function.hpp"
   
   int main()
   {
    isobus::NAME myNAME(0); // Create an empty NAME

    // Set up NAME fields
    myNAME.set_arbitrary_address_capable(true);
    myNAME.set_industry_group(1);
    myNAME.set_device_class(0);
    myNAME.set_function_code(static_cast<std::uint8_t>(isobus::NAME::Function::SteeringControl));
    myNAME.set_identity_number(2);
    myNAME.set_ecu_instance(0);
    myNAME.set_function_instance(0);
    myNAME.set_device_class_instance(0);
    myNAME.set_manufacturer_code(64);

    return 0;
   }

> So now, we have a NAME with actual information encoded in it. In this example, we are an on-highway steering controller, our serial number is 2, we support address arbitration, and we claim to be manufacturerd by manufacturer 64, which is currently an unused code (but was formerly "SpectraPhysics").

Reviewer: ```suggestion
So now, we have a NAME with actual information encoded in it. In this example, we are an on-highway steering controller, our serial number is 2, we support address arbitration, and we claim to be manufacturer by manufacturer 64, which is currently an unused code (but was formerly "SpectraPhysics").
```

> Make sure to change these values to match your device's identity. Use `"isobus.net" <https://www.isobus.net/isobus/>`_ as a resouce if you need to figure out all these values.\

Reviewer: ```suggestion
Make sure to change these values to match your device's identity. Use `"isobus.net" <https://www.isobus.net/isobus/>`_ as a resource if you need to figure out all these values.\
```
